jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java

```diff
@@ -306,6 +332,39 @@ protected void renderGraph(Graphics2D g2d) {
 
     g2d.setTransform(newXform);
 
+    Spatial spatial = ((ObservableCachingLayout) layout).getSpatial();
+
+    AnnotationPaintable lowerAnnotationPaintable = null;
+    // when logging is set to trace, the grid will be drawn on the graph visualization
+    if (log.isTraceEnabled()) {
+      AnnotationRenderer annotationRenderer = new AnnotationRenderer();
+      lowerAnnotationPaintable = new AnnotationPaintable(renderContext, annotationRenderer);
+      if (spatial != null) {
+        Collection<Rectangle2D> grid = ((SpatialGrid) spatial).getGrid();
+        int num = 0;
+        for (Rectangle2D r : grid) {
+          Point2D p =
+              this.getRenderContext()
+                  .getMultiLayerTransformer()
+                  .inverseTransform(new Point2D.Double(r.getX(), r.getY()));
+          Annotation<Shape> annotation =
+              new Annotation<Shape>(r, Annotation.Layer.LOWER, Color.BLACK, false, p);
+          lowerAnnotationPaintable.add(annotation);
+          String label = num + ":" + spatial.getMap().get(num);
+          int stringWidth = g2d.getFontMetrics().stringWidth(label);
+          Point2D center =
+              new Point2D.Double(
+                  r.getX() + (r.getWidth() - stringWidth) / 2, r.getY() + r.getHeight() / 2);
+
+          Annotation<String> annotation2 =
+              new Annotation<String>(label, Annotation.Layer.LOWER, Color.BLACK, false, center);
+          lowerAnnotationPaintable.add(annotation2);
+          num++;
+        }
+        this.addPreRenderPaintable(lowerAnnotationPaintable);
+      }
+    }
```

Comment:
Its not a feature for users, it is very low level debugging information.

Reply:
Did you run the code or watch the video to see it work? I updated the code (not the video) so that each cell also now shows a list of contained nodes. When things are resized and nodes or cells move, the lists update. It would be pretty tedious to debug stuff like that without having a 'picture'.

Reply:
Okay, I see now.

I did see you posted a video over at https://github.com/jrtom/jung/issues/124#issuecomment-335025999 a few days ago, but for some reason I had trouble getting it to start playing on Chrome. :/

However, I've tried Firefox just now, and that seems to work! Seeing the video, I can definitely see that it would be difficult to debug a spatial grid without such a picture.